src/parser/type_checking.c

```diff
@@ -13048,1202 +13190,358 @@ pt_eval_function_type (PARSER_CONTEXT * parser, PT_NODE * node)
+    case PT_PERCENTILE_CONT:
+    case PT_PERCENTILE_DISC:
+    case PT_GROUPBY_NUM:
+    case PT_ROW_NUMBER:
+    case PT_RANK:
+    case PT_DENSE_RANK:
+    case PT_AGG_BIT_AND:
+    case PT_AGG_BIT_OR:
+    case PT_AGG_BIT_XOR:
     case PT_STDDEV:
     case PT_STDDEV_POP:
     case PT_STDDEV_SAMP:
     case PT_VARIANCE:
     case PT_VAR_POP:
     case PT_VAR_SAMP:
     case PT_AVG:
-      if (arg_type != PT_TYPE_MAYBE && arg_type != PT_TYPE_NULL && arg_type != PT_TYPE_NA)
-	{
-	  if (!PT_IS_NUMERIC_TYPE (arg_type))
-	    {
-	      arg_list = pt_wrap_with_cast_op (parser, arg_list, PT_TYPE_DOUBLE, TP_FLOATING_PRECISION_VALUE, 0, NULL);
-	      if (arg_list == NULL)
-		{
-		  return node;
-		}
-	    }
-	  node->info.function.arg_list = arg_list;
-	}
-
-      arg_type = PT_TYPE_DOUBLE;
-      break;
-
-    case PT_AGG_BIT_AND:
-    case PT_AGG_BIT_OR:
-    case PT_AGG_BIT_XOR:
-      if (!PT_IS_DISCRETE_NUMBER_TYPE (arg_type) && arg_type != PT_TYPE_MAYBE && arg_type != PT_TYPE_NULL
-	  && arg_type != PT_TYPE_NA)
+    case PT_COUNT_STAR:
+    case PT_COUNT:
+#if 0//to be moved in constant folding
+      /* do special constant folding; COUNT(1), COUNT(?), COUNT(:x), ... -> COUNT(*) */
+      /* TODO does this belong to type checking or constant folding? */
+      if (pt_is_const (arg_list))
 	{
-	  /* cast arg_list to bigint */
-	  arg_list = pt_wrap_with_cast_op (parser, arg_list, PT_TYPE_BIGINT, 0, 0, NULL);
-	  if (arg_list == NULL)
+	  PT_MISC_TYPE all_or_distinct;
+	  all_or_distinct = node->info.function.all_or_distinct;
+	  if (fcode == PT_COUNT && all_or_distinct != PT_DISTINCT)
 	    {
-	      return node;
+	      fcode = node->info.function.function_type = PT_COUNT_STAR;
+	      parser_free_tree (parser, arg_list);
+	      arg_list = node->info.function.arg_list = NULL;
 	    }
-	  arg_type = PT_TYPE_BIGINT;
-	  node->info.function.arg_list = arg_list;
 	}
+      node->type_enum = PT_TYPE_INTEGER;
       break;
-
+#endif
     case PT_SUM:
-      if (!PT_IS_NUMERIC_TYPE (arg_type) && arg_type != PT_TYPE_MAYBE && arg_type != PT_TYPE_NULL
-	  && arg_type != PT_TYPE_NA && !pt_is_set_type (arg_list))
-	{
-	  /* To display the sum as integer and not scientific */
-	  PT_TYPE_ENUM cast_type = (arg_type == PT_TYPE_ENUMERATION ? PT_TYPE_INTEGER : PT_TYPE_DOUBLE);
-
-	  /* cast arg_list to double or integer */
-	  arg_list = pt_wrap_with_cast_op (parser, arg_list, cast_type, 0, 0, NULL);
-	  if (arg_list == NULL)
-	    {
-	      return node;
-	    }
-	  arg_type = cast_type;
-	  node->info.function.arg_list = arg_list;
-	}
+    case PT_LEAD:
+    case PT_LAG:
+    case PT_CUME_DIST:
+    case PT_PERCENT_RANK:
+    case PT_NTILE:
+    case F_JSON_OBJECT:
+    case F_JSON_ARRAY:
+    case F_JSON_MERGE:
+    case F_JSON_REPLACE:
+    case F_JSON_SET:
+    case F_JSON_ARRAY_APPEND:
+    case F_JSON_INSERT:
+    case F_JSON_REMOVE:
+    case F_JSON_GET_ALL_PATHS:
+    case F_JSON_KEYS:
+    case F_ELT:
+    case F_INSERT_SUBSTRING:
+      {
```

Comment:
Is it possible to integrate the code deciding the return type into the matching/apply signature function ?

Reply:
done for both node->type_enum & node->data_type; it will be in the next commit/merge